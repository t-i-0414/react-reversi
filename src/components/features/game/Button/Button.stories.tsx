import { Meta, StoryObj } from '@storybook/react';
import React from 'react';
import { Button } from './Button';

const meta: Meta<typeof Button> = {
  title: 'features/game/Button',
  component: Button,
  argTypes: {
    text: {
      control: {
        type: 'text',
      },
    },
    type: {
      control: {
        type: 'select',
        options: {
          submit: 'submit',
          button: 'button',
          reset: 'reset',
        },
      },
    },
    onClick: {
      action: 'onReset',
    },
    dataCy: {
      control: {
        type: 'text',
      },
    },
  },
  args: {
    text: 'Game Start',
    type: 'submit',
    dataCy: 'dataCy',
  },
};

export default meta;
type Story = StoryObj<typeof Button>;

export const Normal: Story = {
  render: args => <Button {...args} />,
};
