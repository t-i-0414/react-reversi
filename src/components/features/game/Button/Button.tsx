import React from 'react';
import styled from 'styled-components';
import { ColorMap, SizeMap } from '~/constants';

export interface Props {
  text: string;
  type: 'submit' | 'button' | 'reset';
  onClick?: () => void;
  dataCy?: string;
  dataTestid?: string;
}

export const Button: React.FC<Props> = ({
  text,
  type,
  onClick,
  dataCy,
  dataTestid,
}) => (
  <StyledBottom
    type={type}
    onClick={onClick}
    data-cy={dataCy}
    data-testid={dataTestid}
  >
    {text}
  </StyledBottom>
);

export const StyledBottom = styled.button`
  position: relative;
  z-index: 0;
  display: block;
  padding: 8px 16px;

  /* margin: 0 auto; */
  overflow: hidden;
  font-size: ${SizeMap.FS_16};
  line-height: 1.2;
  background: none;
  border: 1px solid ${ColorMap.BD_BLACK};
  border-radius: 4px;

  &::before {
    position: absolute;
    top: 0;
    left: 0;
    z-index: -2;
    width: 100%;
    height: 100%;
    content: '';
    background-color: ${ColorMap.BG_WHITE};
  }

  &::after {
    position: absolute;
    top: 0;
    left: -100%;
    z-index: -1;
    width: 100%;
    height: 100%;
    content: '';
    background-color: ${ColorMap.BG_BLACK};
    transition: all 0.3s ease-out;
  }

  &:hover {
    color: ${ColorMap.TX_WHITE};
    background: none;

    &::after {
      left: 0;
    }
  }
`;
