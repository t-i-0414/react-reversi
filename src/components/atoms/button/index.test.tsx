import { expect } from '@jest/globals';
import React from 'react';
import renderer from 'react-test-renderer';
import Button from '.';

describe('button unit test', () => {
  it('snapshot', () => {
    const component = renderer.create(
      <Button
        onClick={() => true}
        text='Game Start'
        type='button'
        dataCy='start'
      />,
    );
    expect(component).toMatchSnapshot();
  });
});
