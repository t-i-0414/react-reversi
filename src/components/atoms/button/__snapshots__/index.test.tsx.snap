// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button unit test snapshot 1`] = `
<button
  className="sc-aXZVg mnGHp"
  data-cy="start"
  onClick={[Function]}
  type="button"
>
  Game Start
</button>
`;
